### PR TITLE
Non replicated merge tree deduplication

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add new setting `non_replicated_deduplication_window` for non-replicated MergeTree inserts deduplication.

### CODE REVIEW

**Comment by alesapin** [comment]
Need to try enabled setting by default

File: src/Storages/MergeTree/MergeTreeDeduplicationLog.h [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+    /// How many entries we have in log
+    size_t entries_count;
+};
+
+/// Simple string-key HashTable with fixed size based on STL containers.
+/// Preserves order using linked list and remove elements
+/// on overflow in FIFO order.
+template <typename V>
+class LimitedOrderedHashMap
```
Comment (nikitamikhaylov) [inline src/Storages/MergeTree/MergeTreeDeduplicationLog.h:29]: LRUCache?

Comment (alesapin) [reply src/Storages/MergeTree/MergeTreeDeduplicationLog.h:29]: But it's not the last recently used map, it's the last recently added map.

**Comment by alesapin** [comment]
At least sever didn't crash. I'll check failures.

**Comment by alesapin** [comment]
Support of deduplication for MergeTree in old syntax is quite complicated because we need to serialize part name, which depends on storage partition key, which depends on metadata snapshot. So I'll just ban it.

**Comment by alesapin** [comment]
Ok, everything looks expected. Integration s3 merge tree test depends on the order. Func tests just insert equal data. It's especially not obvious for `01197_summing_enum` test.

**Comment by alesapin** [comment]
I'll disable setting by default and ban for old-syntax merge tree.

**Review by nikitamikhaylov — APPROVED** [review]

**Comment by alesapin** [comment]
need to support alter setting

**Comment by alesapin** [comment]
Let's wait for performance, but I'm almost sure it's not affected.

**Comment by alesapin** [comment]
Yes, no related changes in performance.

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-8347.

**Pull request merged**